File: Level 2 - Core Interview Pack/heap/core_problems.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <unordered_map>
#include <algorithm>
using namespace std;

/**
 * Core Heap Problems in C++
 * Contains essential heap-based algorithms
 */

// Problem 1: Find Kth Largest Element
// Time Complexity: O(n log k)
int findKthLargest(vector<int>& nums, int k) {
    // Min heap of size k
    priority_queue<int, vector<int>, greater<int>> minHeap;
    
    for (int num : nums) {
        minHeap.push(num);
        if (minHeap.size() > k) {
            minHeap.pop();
        }
    }
    
    return minHeap.top();
}

// Problem 2: Heap Sort
void heapify(vector<int>& arr, int n, int i) {
    int largest = i;
    int left = 2 * i + 1;
    int right = 2 * i + 2;
    
    if (left < n && arr[left] > arr[largest])
        largest = left;
    
    if (right < n && arr[right] > arr[largest])
        largest = right;
    
    if (largest != i) {
        swap(arr[i], arr[largest]);
        heapify(arr, n, largest);
    }
}

void heapSort(vector<int>& arr) {
    int n = arr.size();
    
    // Build max heap
    for (int i = n / 2 - 1; i >= 0; i--)
        heapify(arr, n, i);
    
    // Extract elements one by one
    for (int i = n - 1; i > 0; i--) {
        swap(arr[0], arr[i]);
        heapify(arr, i, 0);
    }
}

// Problem 3: Top K Frequent Elements
vector<int> topKFrequent(vector<int>& nums, int k) {
    // Count frequencies
    unordered_map<int, int> freqMap;
    for (int num : nums) {
        freqMap[num]++;
    }
    
    // Min heap based on frequency
    auto comp = [](pair<int, int> a, pair<int, int> b) {
        return a.second > b.second;
    };
    priority_queue<pair<int, int>, vector<pair<int, int>>, decltype(comp)> minHeap(comp);
    
    for (auto& entry : freqMap) {
        minHeap.push(entry);
        if (minHeap.size() > k) {
            minHeap.pop();
        }
    }
    
    // Extract results
    vector<int> result;
    while (!minHeap.empty()) {
        result.push_back(minHeap.top().first);
        minHeap.pop();
    }
    
    return result;
}

// Problem 4: Merge K Sorted Arrays
struct ArrayPointer {
    int arrayIndex;
    int elementIndex;
    int value;
    
    ArrayPointer(int ai, int ei, int v) : arrayIndex(ai), elementIndex(ei), value(v) {}
    
    bool operator>(const ArrayPointer& other) const {
        return value > other.value;
    }
};

vector<int> mergeKSortedArrays(vector<vector<int>>& arrays) {
    vector<int> result;
    priority_queue<ArrayPointer, vector<ArrayPointer>, greater<ArrayPointer>> minHeap;
    
    // Add first element from each array
    for (int i = 0; i < arrays.size(); i++) {
        if (!arrays[i].empty()) {
            minHeap.push(ArrayPointer(i, 0, arrays[i][0]));
        }
    }
    
    // Process heap
    while (!minHeap.empty()) {
        ArrayPointer current = minHeap.top();
        minHeap.pop();
        result.push_back(current.value);
        
        // Add next element from same array
        if (current.elementIndex + 1 < arrays[current.arrayIndex].size()) {
            int nextIdx = current.elementIndex + 1;
            int nextVal = arrays[current.arrayIndex][nextIdx];
            minHeap.push(ArrayPointer(current.arrayIndex, nextIdx, nextVal));
        }
    }
    
    return result;
}

// Problem 5: Find Median from Data Stream
class MedianFinder {
private:
    priority_queue<int> maxHeap; // Lower half
    priority_queue<int, vector<int>, greater<int>> minHeap; // Upper half
    
public:
    MedianFinder() {}
    
    void addNum(int num) {
        maxHeap.push(num);
        minHeap.push(maxHeap.top());
        maxHeap.pop();
        
        if (maxHeap.size() < minHeap.size()) {
            maxHeap.push(minHeap.top());
            minHeap.pop();
        }
    }
    
    double findMedian() {
        if (maxHeap.size() == minHeap.size()) {
            return (maxHeap.top() + minHeap.top()) / 2.0;
        }
        return maxHeap.top();
    }
};

// Problem 6: Last Stone Weight
int lastStoneWeight(vector<int>& stones) {
    priority_queue<int> maxHeap;
    
    for (int stone : stones) {
        maxHeap.push(stone);
    }
    
    while (maxHeap.size() > 1) {
        int stone1 = maxHeap.top(); maxHeap.pop();
        int stone2 = maxHeap.top(); maxHeap.pop();
        
        if (stone1 != stone2) {
            maxHeap.push(stone1 - stone2);
        }
    }
    
    return maxHeap.empty() ? 0 : maxHeap.top();
}

// Helper function to display vector
void displayVector(const vector<int>& vec) {
    cout << "[ ";
    for (int val : vec) {
        cout << val << " ";
    }
    cout << "]" << endl;
}

int main() {
    cout << "=== Core Heap Problems ===" << endl << endl;
    
    // Test 1: Kth Largest Element
    cout << "Test 1: Kth Largest Element" << endl;
    vector<int> nums1 = {3, 2, 1, 5, 6, 4};
    int k1 = 2;
    cout << "Array: "; displayVector(nums1);
    cout << k1 << "th largest: " << findKthLargest(nums1, k1) << endl << endl;
    
    // Test 2: Heap Sort
    cout << "Test 2: Heap Sort" << endl;
    vector<int> nums2 = {12, 11, 13, 5, 6, 7};
    cout << "Before: "; displayVector(nums2);
    heapSort(nums2);
    cout << "After: "; displayVector(nums2);
    cout << endl;
    
    // Test 3: Top K Frequent
    cout << "Test 3: Top K Frequent Elements" << endl;
    vector<int> nums3 = {1, 1, 1, 2, 2, 3};
    int k3 = 2;
    cout << "Array: "; displayVector(nums3);
    vector<int> topK = topKFrequent(nums3, k3);
    cout << "Top " << k3 << " frequent: "; displayVector(topK);
    cout << endl;
    
    // Test 4: Merge K Sorted Arrays
    cout << "Test 4: Merge K Sorted Arrays" << endl;
    vector<vector<int>> arrays = {
        {1, 4, 7},
        {2, 5, 8},
        {3, 6, 9}
    };
    vector<int> merged = mergeKSortedArrays(arrays);
    cout << "Merged: "; displayVector(merged);
    cout << endl;
    
    // Test 5: Median Finder
    cout << "Test 5: Find Median from Data Stream" << endl;
    MedianFinder mf;
    mf.addNum(1);
    mf.addNum(2);
    cout << "Median after adding 1, 2: " << mf.findMedian() << endl;
    mf.addNum(3);
    cout << "Median after adding 3: " << mf.findMedian() << endl << endl;
    
    // Test 6: Last Stone Weight
    cout << "Test 6: Last Stone Weight" << endl;
    vector<int> stones = {2, 7, 4, 1, 8, 1};
    cout << "Stones: "; displayVector(stones);
    cout << "Last stone weight: " << lastStoneWeight(stones) << endl;
    
    return 0;
}

File: heap/core_problems.java
package heap;

/**
 * Core Heap Problems in Java
 * Contains essential heap-based algorithms
 */

import java.util.*;

public class core_problems {

    // Problem 1: Find Kth Largest Element
    // Time Complexity: O(n log k)
    // Space Complexity: O(k)
    static int findKthLargest(int[] nums, int k) {
        // Use min heap of size k
        PriorityQueue<Integer> minHeap = new PriorityQueue<>();

        for (int num : nums) {
            minHeap.offer(num);
            if (minHeap.size() > k) {
                minHeap.poll();
            }
        }

        return minHeap.peek();
    }

    // Problem 2: Heap Sort
    // Time Complexity: O(n log n)
    // Space Complexity: O(1)
    static void heapSort(int[] arr) {
        int n = arr.length;

        // Build max heap
        for (int i = n / 2 - 1; i >= 0; i--) {
            heapify(arr, n, i);
        }

        // Extract elements one by one
        for (int i = n - 1; i > 0; i--) {
            // Move current root to end
            int temp = arr[0];
            arr[0] = arr[i];
            arr[i] = temp;

            // Heapify reduced heap
            heapify(arr, i, 0);
        }
    }

    private static void heapify(int[] arr, int n, int i) {
        int largest = i;
        int left = 2 * i + 1;
        int right = 2 * i + 2;

        if (left < n && arr[left] > arr[largest])
            largest = left;

        if (right < n && arr[right] > arr[largest])
            largest = right;

        if (largest != i) {
            int swap = arr[i];
            arr[i] = arr[largest];
            arr[largest] = swap;

            heapify(arr, n, largest);
        }
    }

    // Problem 3: Top K Frequent Elements
    // Time Complexity: O(n log k)
    // Space Complexity: O(n)
    static int[] topKFrequent(int[] nums, int k) {
        // Count frequencies
        Map<Integer, Integer> freqMap = new HashMap<>();
        for (int num : nums) {
            freqMap.put(num, freqMap.getOrDefault(num, 0) + 1);
        }

        // Min heap based on frequency
        PriorityQueue<Map.Entry<Integer, Integer>> minHeap = new PriorityQueue<>((a, b) -> a.getValue() - b.getValue());

        for (Map.Entry<Integer, Integer> entry : freqMap.entrySet()) {
            minHeap.offer(entry);
            if (minHeap.size() > k) {
                minHeap.poll();
            }
        }

        // Extract results
        int[] result = new int[k];
        int i = 0;
        while (!minHeap.isEmpty()) {
            result[i++] = minHeap.poll().getKey();
        }

        return result;
    }

    // Problem 4: Merge K Sorted Lists (using arrays)
    // Time Complexity: O(N log k) where N is total elements
    // Space Complexity: O(k)
    static class ArrayPointer {
        int arrayIndex;
        int elementIndex;
        int value;

        ArrayPointer(int arrayIndex, int elementIndex, int value) {
            this.arrayIndex = arrayIndex;
            this.elementIndex = elementIndex;
            this.value = value;
        }
    }

    static List<Integer> mergeKSortedArrays(int[][] arrays) {
        List<Integer> result = new ArrayList<>();
        PriorityQueue<ArrayPointer> minHeap = new PriorityQueue<>((a, b) -> a.value - b.value);

        // Add first element from each array
        for (int i = 0; i < arrays.length; i++) {
            if (arrays[i].length > 0) {
                minHeap.offer(new ArrayPointer(i, 0, arrays[i][0]));
            }
        }

        // Process heap
        while (!minHeap.isEmpty()) {
            ArrayPointer current = minHeap.poll();
            result.add(current.value);

            // Add next element from same array
            if (current.elementIndex + 1 < arrays[current.arrayIndex].length) {
                int nextIdx = current.elementIndex + 1;
                int nextVal = arrays[current.arrayIndex][nextIdx];
                minHeap.offer(new ArrayPointer(current.arrayIndex, nextIdx, nextVal));
            }
        }

        return result;
    }

    // Problem 5: Find Median from Data Stream
    static class MedianFinder {
        PriorityQueue<Integer> maxHeap; // Lower half
        PriorityQueue<Integer> minHeap; // Upper half

        public MedianFinder() {
            maxHeap = new PriorityQueue<>((a, b) -> b - a);
            minHeap = new PriorityQueue<>();
        }

        // Time Complexity: O(log n)
        public void addNum(int num) {
            // Add to max heap first
            maxHeap.offer(num);

            // Balance: move largest from max heap to min heap
            minHeap.offer(maxHeap.poll());

            // Ensure max heap has equal or one more element
            if (maxHeap.size() < minHeap.size()) {
                maxHeap.offer(minHeap.poll());
            }
        }

        // Time Complexity: O(1)
        public double findMedian() {
            if (maxHeap.size() == minHeap.size()) {
                return (maxHeap.peek() + minHeap.peek()) / 2.0;
            }
            return maxHeap.peek();
        }
    }

    // Problem 6: Last Stone Weight
    // Time Complexity: O(n log n)
    static int lastStoneWeight(int[] stones) {
        PriorityQueue<Integer> maxHeap = new PriorityQueue<>((a, b) -> b - a);

        for (int stone : stones) {
            maxHeap.offer(stone);
        }

        while (maxHeap.size() > 1) {
            int stone1 = maxHeap.poll();
            int stone2 = maxHeap.poll();

            if (stone1 != stone2) {
                maxHeap.offer(stone1 - stone2);
            }
        }

        return maxHeap.isEmpty() ? 0 : maxHeap.peek();
    }

    // Display array
    static void displayArray(int[] arr) {
        System.out.print("Array: ");
        for (int val : arr) {
            System.out.print(val + " ");
        }
        System.out.println();
    }

    public static void main(String[] args) {
        System.out.println("=== Core Heap Problems ===\n");

        // Test 1: Kth Largest Element
        System.out.println("Test 1: Kth Largest Element");
        int[] nums1 = { 3, 2, 1, 5, 6, 4 };
        int k1 = 2;
        System.out.println("Array: " + Arrays.toString(nums1));
        System.out.println(k1 + "th largest: " + findKthLargest(nums1, k1));
        System.out.println();

        // Test 2: Heap Sort
        System.out.println("Test 2: Heap Sort");
        int[] nums2 = { 12, 11, 13, 5, 6, 7 };
        System.out.println("Before: " + Arrays.toString(nums2));
        heapSort(nums2);
        System.out.println("After: " + Arrays.toString(nums2));
        System.out.println();

        // Test 3: Top K Frequent
        System.out.println("Test 3: Top K Frequent Elements");
        int[] nums3 = { 1, 1, 1, 2, 2, 3 };
        int k3 = 2;
        System.out.println("Array: " + Arrays.toString(nums3));
        System.out.println("Top " + k3 + " frequent: " + Arrays.toString(topKFrequent(nums3, k3)));
        System.out.println();

        // Test 4: Merge K Sorted Arrays
        System.out.println("Test 4: Merge K Sorted Arrays");
        int[][] arrays = {
                { 1, 4, 7 },
                { 2, 5, 8 },
                { 3, 6, 9 }
        };
        System.out.println("Merged: " + mergeKSortedArrays(arrays));
        System.out.println();

        // Test 5: Median Finder
        System.out.println("Test 5: Find Median from Data Stream");
        MedianFinder mf = new MedianFinder();
        mf.addNum(1);
        mf.addNum(2);
        System.out.println("Median after adding 1, 2: " + mf.findMedian());
        mf.addNum(3);
        System.out.println("Median after adding 3: " + mf.findMedian());
        System.out.println();

        // Test 6: Last Stone Weight
        System.out.println("Test 6: Last Stone Weight");
        int[] stones = { 2, 7, 4, 1, 8, 1 };
        System.out.println("Stones: " + Arrays.toString(stones));
        System.out.println("Last stone weight: " + lastStoneWeight(stones));
    }
}
